File: common/src/main/java/juuxel/adorn/block/KitchenCounterBlock.java
package juuxel.adorn.block;

import juuxel.adorn.block.property.FrontConnection;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockBehaviour.Properties;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.EnumProperty;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.ScheduledTickAccess;

public class KitchenCounterBlock extends AbstractKitchenCounterBlock implements BlockWithDescription {
    public static final EnumProperty<FrontConnection> FRONT = EnumProperty.create("front", FrontConnection.class);
    private static final String DESCRIPTION_KEY = "block.adorn.kitchen_counter.description";

    public KitchenCounterBlock(Properties settings) {
        super(settings);
        registerDefaultState(defaultBlockState().setValue(FRONT, FrontConnection.NONE));
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        super.createBlockStateDefinition(builder);
        builder.add(FRONT);
    }

    @Override
    public BlockState getStateForPlacement(BlockPlaceContext ctx) {
        return updateConnections(super.getStateForPlacement(ctx), ctx.getLevel(), ctx.getClickedPos());
    }

    @Override
    protected BlockState updateShape(BlockState state, LevelReader world, ScheduledTickAccess tickView, BlockPos pos, Direction direction, BlockPos neighborPos, BlockState neighborState, RandomSource random) {
        return updateConnections(state, world, pos);
    }

    private BlockState updateConnections(BlockState state, BlockGetter world, BlockPos pos) {
        var facing = state.getValue(FACING);
        var frontState = world.getBlockState(pos.relative(facing));
        var frontConnection = FrontConnection.NONE;

        if (frontState.getBlock() instanceof AbstractKitchenCounterBlock) {
            var frontFacing = frontState.getValue(FACING);
            if (frontFacing == facing.getClockWise()) {
                frontConnection = FrontConnection.LEFT;
            } else if (frontFacing == facing.getCounterClockWise()) {
                frontConnection = FrontConnection.RIGHT;
            }
        }

        return state.setValue(FRONT, frontConnection);
    }

    @Override
    public String getDescriptionKey() {
        return DESCRIPTION_KEY;
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/KitchenCounterBlock.kt
package juuxel.adorn.block

import juuxel.adorn.api.block.BlockVariant
import juuxel.adorn.block.property.FrontConnection
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.item.ItemPlacementContext
import net.minecraft.state.StateManager
import net.minecraft.state.property.EnumProperty
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.WorldAccess

open class KitchenCounterBlock(variant: BlockVariant) : AbstractKitchenCounterBlock(variant), BlockWithDescription {
    override val descriptionKey = "block.adorn.kitchen_counter.description"

    init {
        defaultState = defaultState.with(FRONT, FrontConnection.NONE)
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        super.appendProperties(builder)
        builder.add(FRONT)
    }

    override fun getPlacementState(context: ItemPlacementContext): BlockState {
        return getStateForNeighborUpdate(
            super.getPlacementState(context),
            null, null, context.world, context.blockPos, null
        )
    }

    override fun getStateForNeighborUpdate(
        state: BlockState, direction: Direction?, neighborState: BlockState?, world: WorldAccess, pos: BlockPos, neighborPos: BlockPos?
    ): BlockState {
        val facing = state[FACING]
        val frontState = world.getBlockState(pos.offset(facing))
        val frontConnection =
            if (frontState.block is AbstractKitchenCounterBlock) {
                when (frontState[FACING]) {
                    facing.rotateYClockwise() -> FrontConnection.LEFT
                    facing.rotateYCounterclockwise() -> FrontConnection.RIGHT
                    else -> FrontConnection.NONE
                }
            } else {
                FrontConnection.NONE
            }

        return state.with(FRONT, frontConnection)
    }

    companion object {
        val FRONT = EnumProperty.of("front", FrontConnection::class.java)
    }
}
